Updates to securesystemslib==0.10.04 (stand-alone)

lukpueh opened the pull request:
Fixes https://github.com/in-toto/in-toto/issues/45
- removes sub-packages `ssl_crypto` and `ssl_commons`
  they were merged into the stand-alone package `securesystemslib`
- updates requirements.txt, requirements-ci.txt and setup.py
  installation of `pycrypto` and `cryptography` is now handled by `securesystemslib`'s  
  requirements files
- updates in_toto demo README.md
- updates import statements
- updates` import_rsakey_from_pem` and `format_rsakey_from_pem`
  to` import_rsakey_from_private_pem` and `import_rsakey_from_public_pem`

**Noteworthy:**
`securesystemslib` used to be part of [`The Update Framework`](https://github.com/theupdateframework/tuf). When we started to implement `in-toto` we sought ways to use the common modules for both projects. We reviewed various approaches/tools and eventually decided to take out the common code from TUF, place it into a separate repository and re-integrate it as submodule in both projects.

This involved replacing absolute imports, e.g. `import (tuf|in_toto).securesystemslib.keys`, with relative imports in order to make `securesystemslib` agnostic to the parent project. Furthermore it forced the maintainers to learn new facets of git, e.g. cloning recursively.

The main **drawbacks** were employing new git commands and not being able to have `securesystemslib` as a stand-alone package and the benefits that come with it, e.g. requirements handling, established test setups, ...).

The **advantage** was that TUF and in_toto could henceforth be released without an additional external dependency, which some platforms (e.g. debian) would require us to vendor, especially in this case where `securesystemslib` is a mostly internal library. Not having to vendor plus the use of git submodules allowed us to easily sync with or report back developments to `securesystemslib`.

The pros and cons led to a conflict between the maintainers of in-toto and TUF. It was decided to create a stand-alone package. Unfortunately, there is no sensible way so that TUF can use `securesystemslib` as stand-alone package and in-toto use it as submodule.
- `git subtree` boils down to hard copying the files with the [feature](http://git.661346.n2.nabble.com/subtree-merges-lose-prefix-after-rebase-td7332850.html) of breaking `git rebase`
- Hard-copying the files would work, if the import statements  inside of `securesystemslib` hadn't been changed to require `securesystemslib` to be on the path (which is normal for stand-alone packages).
- If we decide to vendor, which we will eventually do, we will have to either patch the path (messy) or the import statements (messy)

Commit:
Updates to securesystemslib==0.10.03 (stand-alone)

- removes subpackages ssl_crypto and ssl_commons
  they were merged into the stand-alone package securesystemslib
- updates requirements.txt, requirements-ci.txt and setup.py
  installation of pycrypto and cryptography is now handled by the
  securesystemslib's requirements files
- updates in_toto demo README
- updates import statements
- updates import_rsakey_from_pem and format_rsakey_from_pem
  to import_rsakey_from_private_pem and
  import_rsakey_from_public_pem

Commit:
Remove securesystemslib settings from settings.py

This commit removes securesystemslib specific settings from the
global settings file.

Before securesystemslib became a stand-alone package it used to
import settings from simple_settings, a package that provided a
single global settings file.
simple_settings support was dropped and securesystemslib went back
to use its own local settings file.

vladimir-v-diaz commented on in_toto/models/common.py:
```diff
@@ -102,6 +102,7 @@ def verify_signatures(self, keys_dict):
       except:
```
We should avoid blank except statements.

vladimir-v-diaz commented on in_toto/models/common.py:
```diff
@@ -102,6 +102,7 @@ def verify_signatures(self, keys_dict):
         raise SignatureVerificationError(
             "Signature key not found, key id is '{0}'".format(keyid))
-      if not keys.verify_signature(key, signature, self.payload):
+      if not securesystemslib.keys.verify_signature(
```
This can potentially raise an exception. For example, `securesystems.exceptions.UnknownMethodError` is raised if `key` specifies an unsupported key type.  I guess this function assumes these arguments are valid, or verified elsewhere? (Perhaps assumptions of these sort should be stated in the docstring.)

lukpueh replied:
That's actually fine. The method is expected to raise an exception if the signature does not verify (for whatever reason).

vladimir-v-diaz replied:
Ahh okay.  You should probably add a doctring specifying which exceptions it raises, arguments, etc.

lukpueh replied:
I am not sure, if the `<Exceptions>` docstring of a function should also list the possible exceptions of the functions it is calling.

Commit:
Fixes blank except statements

vladimir-v-diaz commented on in_toto/matchrule_validators.py:
```diff
@@ -35,40 +36,47 @@ def _validate_match_rule(keywords):
   if not isinstance(keywords, list):
-    raise FormatError("this matching rule is not a list")
+    raise securesystemslib.exceptions.FormatError(
+        "this matching rule is not a list")
```
I find it helpful if the error message includes the argument/value that is wrong.  When appropriate, examples of valid values also helps.

The following error message, used later in the code, is good:
```Python
raise securesystemslib.exceptions.FormatError("{} is not a valid rule!".format(rule))
```

lukpueh replied:
Agreed.

vladimir-v-diaz commented on in_toto/exceptions.py:
```diff
@@ -1,4 +1,4 @@
-from in_toto.ssl_commons.exceptions import Error
+from securesystemslib.exceptions import Error
```
```Python
class RuleVerficationFailed(Error):
```
The exception name here is inconsistent with the others.  Should it instead be `RuleVerificationFailedError`?

lukpueh replied:
Fixed in 2f39a18

vladimir-v-diaz commented on in_toto/models/common.py:
```diff
@@ -99,9 +99,10 @@ def verify_signatures(self, keys_dict):
       try:
         key = keys_dict[keyid]
-      except:
+      except KeyError:
```
This method doesn't validate it's arguments, so I'm not sure where/when/how its done, but `keys_dict[keyid]` can potentially raise `TypeError` if `keyid` happens to contain an unhashable type.

```Python
>>> dict1 = {}
>>> dict1[{'foo': 1}]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unhashable type: 'dict'
```

lukpueh replied:
Fixed in ed60dd7

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -245,45 +253,51 @@ def read(data):
       pass
-    except:
-      raise FormatError("Invalid threshold value for this step")
+    except Exception:
```
Is this incomplete or a WIP?

```Python
def _validate_threshold(self):
    """Private method to check that the threshold field is set to an int."""
    try:
      # int(self.threshold)
      # FIXME: we don't have support for threshold yet
      pass
    except Exception:
      raise securesystemslib.exceptions.FormatError(
"Invalid threshold value for this step")
```

lukpueh replied:
This is WIP. @RohanBhirangi is working on adding the threshold field.

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -245,45 +253,51 @@ def read(data):
-    except:
-      raise FormatError("Invalid threshold value for this step")
+    except Exception:
+      raise securesystemslib.exceptions.FormatError(
```
Debugging becomes difficult when you catch `Exception` and raise your own without logging or printing the original exception.

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -320,25 +334,29 @@ def read(data):
     if self._type != "inspection":
-      raise FormatError("The _type field should be aset to inspection!")
+      raise securesystemslib.exceptions.FormatError(
+          "The _type field should be aset to inspection!")
```
typo

vladimir-v-diaz commented on requirements-ci.txt:
```diff
@@ -1,5 +1,3 @@
-pycrypto
-cryptography
 pytest
 PyNaCl
```
PyNaCl is also a dependency of the securesystemslib library.

vladimir-v-diaz commented:
I do not see any other issues with the securesystemslib bits.

vladimir-v-diaz commented on in_toto/models/layout.py:
```diff
@@ -148,58 +147,67 @@ def import_step_metadata_from_files_as_dict(self):
       if 'private' in key and key['private'] != '':
-        raise FormatError("key: {} contains a private key part!".format(keyid))
+        raise securesystemslib.exceptions.FormatError(
+            "key: {} contains a private key part!".format(keyid))
```
v0.10.4 now contains the schema for public keyvals:
https://github.com/secure-systems-lab/securesystemslib/releases/tag/v0.10.4
https://pypi.python.org/pypi?:action=display&name=securesystemslib&version=0.10.4

lukpueh replied:
I updated to 0.10.4 in https://github.com/in-toto/in-toto/pull/65/commits/df25016ff32603935788b5a662d648a94256609e.
Unfortunately I can't rename the branch without loosing this GitHub discussion but I updated the PR title.

Commit:
Include argument in validator exceptions

Commit:
Rename exception Failed to Error for conistency

Commit:
Adds key format check to signature methods

Commit:
Adds original exception message in re-raise

Commit:
Fix typo in exception message

Commit:
Remove securesystemslib requirement PyNaCl

Commit:
Updates to securesystemslib v0.10.04

The new version includes schemas for PUBLIC_KEY and PUBLIC_KEYVAL

Commit:
Uses PUBLIC_KEY_SCHEMA in layout, util and tests

Uses the new securesystemlib schema PUBLIC_KEY_SCHEMA to guarantee
that a public RSA key does not have a private portion

lukpueh commented:
@vladimir-v-diaz, thanks for the review! I added your suggestion to the big [user feedback/exception messages issue](https://github.com/in-toto/in-toto/issues/6#issuecomment-274845182), because user it should be improved all over the place.

vladimir-v-diaz commented on in_toto/matchrule_validators.py:
```diff
@@ -25,7 +25,8 @@ def _validate_match_rule(keywords):
   if not isinstance(keywords, list):
     raise securesystemslib.exceptions.FormatError(
-        "this matching rule is not a list")
+        "Wrong rule format, must be a list.\n\t"
```
Not part of this PR, but the <Module Name> entry in the header block is wrong.

vladimir-v-diaz replied:
Also not part of this PR, but our lab projects should choose a single way to wrap long lines.

I spoke with @awwad about this, but we think indenting wrapped lines looks best.  For example,

```Python
raise FormatError("Argument"
  " is improperly formatted.  Valid"
  " format of argument is X.")
```
Note, that's 2 spaces per indentation.

I've used the following in the past, but it's more difficult to read and edit:

```Python
raise FormatError("Argument is improperly"
                  " formatted.  Valid format"
                  " of argument is X.")
```

lukpueh replied:
Re header block: Thanks, will fix.
Re indentation: I operate according to our code style guide lines here, which suggests [double-indentation for line continuation](https://github.com/secure-systems-lab/code-style-guidelines#maximum-line-length)

vladimir-v-diaz replied:
That's correct, I forgot that we decided to use double-indentation.

Commit:
Fixes module docstring in matchrule_validators

vladimir-v-diaz commented on in_toto/matchrule_validators.py:
```diff
@@ -1,6 +1,6 @@
 """
 <Module Name>
-  in_toto/models/validators.py
+  in_toto/models/matchrule_validators.py
```
I don't find a `matchrule_validators.py` in the `in-toto/models` directory.

lukpueh replied:
Sorry! `git --amend` to the rescue. :)

vladimir-v-diaz commented on in_toto/matchrule_validators.py:
```diff
@@ -37,46 +38,51 @@ def _validate_match_rule(keywords):
-    raise securesystemslib.exceptions.FormatError(
-        "Wrong rule to verify! {}".format(rule))
+    raise securesystemslib.exceptions.FormatError("Wrong rule format.\n\t"
+        "Got: {}".format(keywords))
```
The user could probably figure it out on their own, but a more helpful exception would emphasize that the rule keyword is wrong (i.e., must be `MATCH` or `match`).

vladimir-v-diaz commented on in_toto/matchrule_validators.py:
```diff
@@ -37,46 +38,51 @@ def _validate_match_rule(keywords):
   if len(keywords) != 2:
-    raise securesystemslib.exceptions.FormatError("Wrong rule format")
+    raise securesystemslib.exceptions.FormatError("Wrong rule format.\n\t"
+        "Got: {}".format(keywords))
```
It would help the user more if the exception message emphasized that len(keywords) must be 2, like you do with the others.

vladimir-v-diaz commented:
Okay, feel free to merge when you're done with the remaining comments.

Commit:
Enhances matchrule validator exception messages
